sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -595,8 +595,58 @@ class Analyzer(
       }
     }
 
+    private def tryResolveHavingCondition(
+        a: UnresolvedHaving, havingCond: Expression, agg: LogicalPlan): LogicalPlan = {
+      val aggForResolving = agg match {
+        case a: Aggregate =>
+          // For CUBE/ROLLUP expressions, to avoid resolving repeatedly, here we delete them from
+          // groupingExpressions for condition resolving.
+          a.copy(groupingExpressions = Seq.empty)
+        case g: GroupingSets =>
+          Aggregate(g.groupByExprs, g.aggregations, g.child)
+      }
+      // Try resolving the condition of the filter as though it is in the aggregate clause
+      val (extraAggExprs, transformedAggregateFilter) =
+        ResolveAggregateFunctions.resolveFilterCondInAggregate(
+          havingCond, aggForResolving, resolveFilterNotInGroupingExprs = true)
+
+      // Push the aggregate expressions into the aggregate (if any).
+      if (extraAggExprs.nonEmpty) {
+        val newChild = agg match {
+          case Aggregate(Seq(c @ Cube(groupByExprs)), aggregateExpressions, child) =>
+            constructAggregate(
+              cubeExprs(groupByExprs), groupByExprs, aggregateExpressions ++ extraAggExprs, child)
+          case Aggregate(Seq(r @ Rollup(groupByExprs)), aggregateExpressions, child) =>
+            constructAggregate(
+              rollupExprs(groupByExprs), groupByExprs, aggregateExpressions ++ extraAggExprs, child)
+          case x: GroupingSets =>
+            constructAggregate(
+              x.selectedGroupByExprs, x.groupByExprs, x.aggregations ++ extraAggExprs, x.child)
+        }
+        Project(newChild.output.filter(_.name != "havingCondition"),
+          Filter(transformedAggregateFilter.get, newChild))
+      } else {
+        a
+      }
+    }
+
     // This require transformUp to replace grouping()/grouping_id() in resolved Filter/Sort
```

Comment:
@xuanyuanking Change this comment to say why we want to do a top-down traversal ?

Reply:
Ah yes, thanks for the reminding, done in f2a2ad1.